File: Queue/Queue.py
import numpy as np

# Class Queue


class Queue:
    """
    size = 100
    dtypes = int, float, string, complex
    """

    def __init__(self, size: int = 100, dtype: str = 'int') -> None:
        self.size = size - 1
        self.dtypes = {'int': int, 'float': float,
                       'string': str, 'complex': complex}
        self.dtype = self.dtypes.get(dtype) if self.dtypes.get(
            dtype) else self.dtypes.get('int')
        self.queue = np.zeros((size,), dtype=self.dtype)
        self.front = -1
        self.rear = -1
        self.free = size - 1

    def enqueue(self, data: int) -> bool:
        if self.free == 0:
            print("Queue Full!")
            return False
        if type(data) != self.dtype:
            print("TypeError!")
            return False
        self.queue[self.rear] = data
        self.free -= 1
        self.rear += 1
        self.rear %= self.size
        return True

    def dequeue(self):
        if self.free == self.size:
            print("Queue is Empty!")
            return None
        data = self.queue[self.front]
        self.front += 1
        self.front %= self.size
        self.free += 1
        return data

    def show(self):
        if self.free == self.size:
            print("Queue is Empty!")
            return
        tempFront, tempRear = self.front, self.rear
        count = 0
        while count < (self.size - self.free):
            print(self.queue[tempFront], end='=>')
            tempFront += 1
            tempFront %= self.size
            count += 1

    def freeSpace(self) -> int:
        return self.free

    def peek(self) -> int:
        return self.queue[self.front]


if __name__ == '__main__':
    queue = Queue(5)
    while True:
        option = input(
            "\nChoose: 1.Enqueue() 2.Dequeue() 3.Peek() 4.isEmpty() 5.free() 6.Show() -1. Exit: ")
        if option == '1':
            data = int(input("Data: "))
            queue.enqueue(data)
        elif option == '2':
            print(f'Poped Data: {queue.dequeue()}')
        elif option == '3':
            print(f'Peek Data: {queue.peek()}')
        elif option == '4':
            print(f'Is Empty: {queue.isEmpty()}')
        elif option == '5':
            print(f'AVailable Space: {queue.freeSpace()}')
        elif option == '6':
            queue.show()
        if option == '-1':
            exit()
        else:
            print("Choose Valid Option")

File: Queue/Queue.cpp
#include <iostream>

// Class
class Queue
{
private:
    int front, rear, size, free;
    int *queue;

public:
    Queue(int size)
    {
        this->front = 0;
        this->rear = -1;
        this->size = size - 1;
        this->free = size - 1;
        this->queue = new int[size - 1];
        std::cout << "\nQueue Initialized!";
    }

    int enqueue(int data)
    {
        if (this->free == 0)
        {
            std::cout << "\n Queue id Full";
            return 0;
        }

        this->rear++;
        this->rear = this->rear % this->size;
        this->queue[this->rear] = data;
        this->free--;
        return 1;
    }

    int dequeue()
    {
        if (this->free == this->size)
        {
            std::cout << "\nQueue is Empty!";
            return -1;
        }
        int data = this->queue[this->front];
        this->front++;
        this->front = this->front % this->size;
        this->free++;
        return data;
    }

    void show()
    {
        int tempFront, count = 0;
        tempFront = this->front;
        while (count < (this->size - this->free))
        {
            std::cout << "=>" << this->queue[tempFront];
            ++tempFront;
            tempFront = tempFront % this->size;
            count++;
        }
    }

    int peek()
    {
        return this->queue[this->front];
    }

    bool isEmpty()
    {
        if (this->size == this->free)
        {
            return 1;
        }
        return 0;
    }

    int freeSpace()
    {
        return this->free;
    }
};

int main(int argc, char *argv[])
{
    int option, data;
    Queue queue = Queue(5);
    do
    {
        std::cout << "\n=======Choose======" << std::endl;
        std::cout << "1.Push || 2.Pop || 3.show || 4.Peek || 5. isEmpy || 6.free : ";
        std::cin >> option;
        switch (option)
        {
        case 1:
            std::cout << "\n Enter Data:";
            std::cin >> data;
            queue.enqueue(data);
            break;
        case 2:
            data = queue.dequeue();
            std::cout << "\n Popped: " << data;
            break;
        case 3:
            queue.show();
            break;
        case 4:
            std::cout << "\nData: " << queue.peek();
            break;
        case 5:
            std::cout << "\nEmpty: " << queue.isEmpty();
            break;
        case 6:
            std::cout << "\nAvailable Space: " << queue.freeSpace();
            break;
        default:
            std::cout << "\nChoose Valid Option";
            break;
        }

    } while (option != 0);
    return 0;
}
